### PR TITLE
set EXPLICIT_OBJECTIDS header

### PR DESCRIPTION
**Pull request by fmoor**
fixes https://github.com/edgedb/edgedb-go/issues/124

### CODE REVIEW

**Comment by fmoor** [comment]
WIP until there is a nightly build that includes https://github.com/edgedb/edgedb/pull/2443

File: connect.go [hunk 1]
```diff
@@ -25,16 +25,51 @@ import (
...
+		return true
+	case v.major < other.minor:
+		return false
+	default:
+		return v.minor > other.minor
+	}
+}
+
+func (v version) gte(other version) bool {
+	if v == other {
```
Comment (1st1) [inline connect.go:50]: Does `==` compares all fields or just the pointers?

File: internal/codecs/nil.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/edgedb/edgedb-go/internal/buff"
+	"github.com/edgedb/edgedb-go/internal/descriptor"
+	types "github.com/edgedb/edgedb-go/internal/edgedbtypes"
+)
+
+type nilDecoder struct{}
+
+func (c nilDecoder) DescriptorID() types.UUID { return descriptor.NilID }
+
+func (c nilDecoder) Decode(r *buff.Reader, out unsafe.Pointer) {}
```
Comment (1st1) [inline internal/codecs/nil.go:31]: Hm, this seems to be something that doesn't decode literal `NULL`/empty sets into `nil`, it's just a codec that does nothing, is that right?

File: granularflow.go [hunk 3]
```diff
@@ -110,9 +111,15 @@ func (c *baseConn) pesimistic(r *buff.Reader, q *gfQuery) error {
 }
 
 func (c *baseConn) prepare(r *buff.Reader, q *gfQuery) (idPair, error) {
+	headers := copyHeaders(q.headers)
+
+	if c.protocolVersion.gte(version{0, 10}) {
```
Comment (1st1) [inline granularflow.go:116]: I think we should set some internal bool flag like `c.ver_0_10` to avoid this function call.

File: granularflow.go [hunk 4]
```diff
@@ -190,9 +197,15 @@ func (c *baseConn) describe(r *buff.Reader, q *gfQuery) (descPair, error) {
 			// input descriptor
 			descs.in = descriptor.Pop(r.PopSlice(r.PopUint32()))
 
-			// output descriptor
-			r.Discard(16) // descriptor ID
-			descs.out = descriptor.Pop(r.PopSlice(r.PopUint32()))
+			// output descriptor ID
+			outID := r.PopUUID()
+
+			if outID == descriptor.NilID {
```
Comment (1st1) [inline granularflow.go:203]: Hm, I'm not sure I understand how this check works... outID comes from our protocol, right?

File: connect.go (hunk 1 above)
Comment (fmoor) [reply connect.go:50]: If `v` and `other` were pointers then `==` would only compare the pointers, but they are both passed by value to this function so the struct fields are compared.

File: internal/codecs/nil.go (hunk 2 above)
Comment (fmoor) [reply internal/codecs/nil.go:31]: Correct it decodes empty blocks corresponding to descriptor id 0. I'm open to other names. I didn't think too hard about that one.

File: granularflow.go (hunk 4 above)
Comment (fmoor) [reply granularflow.go:203]: correct. from the [docs](https://www.edgedb.com/docs/internals/protocol/typedesc#type-descriptors):

> There is one special type with type id of zero: 00000000-0000-0000-0000-000000000000. The describe result of this type contains zero blocks. It’s used when a statement returns no meaningful results, e.g. the CREATE DATABASE example statement.

Comment (fmoor) [reply granularflow.go:203]: maybe I should call it `descriptor.ZeroID` :thinking: then `nilDecoder` would become `zeroDecoder`... or something.

File: internal/codecs/nil.go (hunk 2 above)
Comment (1st1) [reply internal/codecs/nil.go:31]: I'd call it "NoOpDecoder"

File: granularflow.go (hunk 4 above)
Comment (1st1) [reply granularflow.go:203]: Ah, I get it now, thanks

**Review by 1st1 — APPROVED** [review]
Approving this so you can commit when the nits are addressed

File: internal/codecs/nil.go (hunk 2 above)
Comment (fmoor) [reply internal/codecs/nil.go]: perfect! I came to the same conclusion independently

changed

**Pull request merged**